rosbag2_transport/src/rosbag2_transport/player_service_client.cpp

```diff
@@ -0,0 +1,318 @@
+// Copyright 2023 Sony Group Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+
+#include "rosbag2_transport/player_service_client.hpp"
+
+#include "rosbag2_cpp/service_utils.hpp"
+
+#include "rosidl_typesupport_introspection_cpp/identifier.hpp"
+#include "rosidl_typesupport_introspection_cpp/message_introspection.hpp"
+
+#include "logging.hpp"
+
+namespace rosbag2_transport
+{
+
+PlayerServiceClient::PlayerServiceClient(
+  std::shared_ptr<rclcpp::GenericClient> cli,
+  std::string service_name,
+  const std::string & service_event_type,
+  const rclcpp::Logger logger,
+  std::shared_ptr<PlayerServiceClientManager> player_service_client_manager)
+: client_(std::move(cli)),
+  service_name_(std::move(service_name)),
+  logger_(logger),
+  player_service_client_manager_(player_service_client_manager)
+{
+  ts_lib_ = rclcpp::get_typesupport_library(
+    service_event_type, "rosidl_typesupport_cpp");
+
+  service_event_type_ts_ = rclcpp::get_message_typesupport_handle(
+    service_event_type,
+    "rosidl_typesupport_cpp",
+    *ts_lib_);
+
+  auto service_event_ts_introspection = get_message_typesupport_handle(
+    service_event_type_ts_,
+    rosidl_typesupport_introspection_cpp::typesupport_identifier);
+
+  message_members_ =
+    reinterpret_cast<const rosidl_typesupport_introspection_cpp::MessageMembers *>(
+    service_event_ts_introspection->data);
+}
+
+bool PlayerServiceClient::include_request_message(
```

Comment:
i think this assumption is the current best effort.

it is likely expected that user enable service introspection on service server side.
not saying this is always, but likely since it is much easier to get the service bag compared to enable all clients on the service.
i believe that user enables client side if they cannot enable it on server side, this constraint could happen with 3rd party driver and node.

btw, this can be much more simple if we resolve https://github.com/ros2/rmw/issues/357 ... lets assign that to me.

Reply:
> i believe that user enables client side if they cannot enable it on server side, this constraint could happen with 3rd party driver and node.

Yes. So I think it is better to provide a way for user to enable client introspection.  
Like timeout and queue length for request future, we should consider how to provide this interface to user.  

> https://github.com/ros2/rmw/issues/357
This must be resolved. Otherwise, while service introspection and client introspection are enable at the same time, rosbag play cannot work on FastDDS and cycleDDS.